builder/lxc/step_lxc_create.go

```diff
@@ -17,51 +19,83 @@ type stepLxcCreate struct{}
+		commands[0] = append(commands[0], config.LxcTemplate.Parameters...)
+		// prevent tmp from being cleaned on boot, we put provisioning scripts there
+		// TODO: wait for init to finish before moving on to provisioning instead of this
+		commands[1] = []string{"touch", filepath.Join(rootfs, "tmp", ".tmpfs")}
+
+		ui.Say("Creating container from template...")
+		for _, command := range commands {
+			log.Printf("Executing sudo command: %#v", command)
+			if err := s.SudoCommand(command...); err != nil {
+				errorHandler(fmt.Errorf("Command (%s) failed with error: %s", strings.Join(command, " "), err))
+				return multistep.ActionHalt
+			}
+		}
 	} else {
-		commands[0] = []string{"lxc-clone", "--orig", config.CloneSource, "--new", name}
-	}
-	// prevent tmp from being cleaned on boot, we put provisioning scripts there
-	// todo: wait for init to finish before moving on to provisioning instead of this
-	commands[1] = []string{"touch", filepath.Join(rootfs, "tmp", ".tmpfs")}
-	commands[2] = []string{"lxc-start", "-d", "--name", name}
-
-	ui.Say("Creating container...")
-	for _, command := range commands {
-		log.Printf("Executing sudo command: %#v", command)
-		err := s.SudoCommand(command...)
+		containerPath := filepath.Join(lxc_dir, name)
+		containerConfig, err := NewLxcConfig(config.RootFs.ConfigFile)
+		if err != nil {
+			errorHandler(fmt.Errorf("Could not read lxc config (%s): %s", config.RootFs.ConfigFile, err))
+			return multistep.ActionHalt
+		}
+		containerConfig.SetRootFs(rootfs)
+		tmpDir, err := ioutil.TempDir("", "lxcconfig")
 		if err != nil {
-			err := fmt.Errorf("Error creating container: %s", err)
-			state.Put("error", err)
-			ui.Error(err.Error())
+			errorHandler(fmt.Errorf("Could not create temp directory (%s): %s", tmpDir, err))
 			return multistep.ActionHalt
 		}
+		err = containerConfig.Write(filepath.Join(tmpDir, "lxc.config"))
+		if err != nil {
+			os.RemoveAll(tmpDir)
+			errorHandler(fmt.Errorf("Could not write lxc config to %s: %s", filepath.Join(tmpDir, "lxc.config"), err))
+			return multistep.ActionHalt
+		}
+
+		commands := make([][]string, 3)
+		commands[0] = []string{"mkdir", containerPath}
+		commands[1] = []string{"tar", "-C", containerPath, "-xf", config.RootFs.Archive}
+		commands[2] = []string{"cp", filepath.Join(tmpDir, "lxc.config"), filepath.Join(containerPath, "config")}
+
+		ui.Say(fmt.Sprintf("Creating container from archive: %s", config.RootFs.Archive))
+		for _, command := range commands {
+			log.Printf("Executing sudo command: %#v", command)
+			if err := s.SudoCommand(command...); err != nil {
+				errorHandler(fmt.Errorf("Command (%s) failed with error: %s", strings.Join(command, " "), err))
+				return multistep.ActionHalt
+			}
+		}
+		os.RemoveAll(tmpDir)
+	}
+	ui.Say("Starting container...")
+	if err := s.SudoCommand("lxc-start", "-d", "-n", name); err != nil {
+		errorHandler(fmt.Errorf("Error starting container: %s", err))
+		return multistep.ActionHalt
 	}
```

Comment:
With long branching logic paths it would be really nice to see the actual LXC creation steps for each to be in their own method. It's starting to get difficult to read and understand `Run` at a glance.